### PR TITLE
Refactor type generator parse method

### LINKED ISSUE #2964: v34.1 UnhandledPromiseRejectionWarning on generating types
After upgrading to v34.1, I started getting the following console error upon generating types:

```
(node:33336) UnhandledPromiseRejectionWarning: SyntaxError: Unexpected token, expected "}" (13:28)      
    at Object._raise (D:\CENGYS\Dev\cengys-r\node_modules\@babel\parser\lib\index.js:776:17)
    at Object.raiseWithData (D:\CENGYS\Dev\cengys-r\node_modules\@babel\parser\lib\index.js:769:17)     
    at Object.raise (D:\CENGYS\Dev\cengys-r\node_modules\@babel\parser\lib\index.js:737:17)
    at Object.unexpected (D:\CENGYS\Dev\cengys-r\node_modules\@babel\parser\lib\index.js:9735:16)       
    at Object.expect (D:\CENGYS\Dev\cengys-r\node_modules\@babel\parser\lib\index.js:9721:28)
    at Object.jsxParseExpressionContainer (D:\CENGYS\Dev\cengys-r\node_modules\@babel\parser\lib\index.js:5028:10)
    at Object.jsxParseElementAt (D:\CENGYS\Dev\cengys-r\node_modules\@babel\parser\lib\index.js:5119:36)    at Object.jsxParseElement (D:\CENGYS\Dev\cengys-r\node_modules\@babel\parser\lib\index.js:5162:17)  
    at Object.parseExprAtom (D:\CENGYS\Dev\cengys-r\node_modules\@babel\parser\lib\index.js:5169:19)    
    at Object.parseExprSubscripts (D:\CENGYS\Dev\cengys-r\node_modules\@babel\parser\lib\index.js:10708:23)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:33336) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not 
handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:33336) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.  
```

I checked, and this is specific to my project, and a simple test project would not generate this error. Unfortunately, no clues were provided in the above log as to which file in my code generated the error, so I added a bunch of console.log statements into the redwood framework until I tracked it down to the following lines of redwoodjs code, which calls babel traverse.
```
export const hasDefaultExport = (code: string): boolean => {
  let exported = false
  const ast = parse(code) as types.Node
    traverse(ast, {
      ExportDefaultDeclaration() {
        exported = true
        return
      },
    })
  return exported
}
```

The offending code in my project was:

```
  const from = await user({ id: <number>submittal.fromId })
  const to = await user({ id: <number>submittal.toId })
```
Note that this code worked in all other respects.  Changing the code to the following resolved the problem.

```
 const from = await user({ id: submittal.fromId as number })
 const to = await user({ id: submittal.toId as number })
```

So, this appears to be a babel traverse issue and not super-critical, as once found, it can easily be fixed.  
However, how could we make things easier for someone to find the offending code that causes babel to puke?  In my case, I had to edit the framework to find the error.  As this appears to be related to asynchronous code in babel, a simple try-catch block in the function does not appear to work.   

The only idea I have is to add a verbose flag to the generate types command so that it spits out filenames to the console as they are processed.

@peterp, any thoughts or suggestions?

### LINKED ISSUE #3351: Please don't swallow graphql code gen errors
These lines swallow really helpful error messages
https://github.com/redwoodjs/redwood/blob/306c26ee558d4a03bb7e234828817763dbf5b4f6/packages/internal/src/generate/graphqlSchema.ts#L54-L62

Reading the comment in the snippet above it sounds like the ambition was to do more. I'd like to follow through on that.

For example, if I have this as part of one of my .sdl files

```sdl
type ProductCategory {
  code: String!
  name: String!
  isEmpty: Bool!
}
```

Currently all I get is this

```
Generating...

Failed to load schema
Failed to load schema

Error: Could not generate GraphQL schema
```

The full error (`e` in the code above) is this

```
ListrError: Something went wrong
    at C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\listr\index.js:102:18
    at async executeCodegen (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\@graphql-codegen\cli\index.cjs.js:934:5)
    at async generate (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\@graphql-codegen\cli\index.cjs.js:1231:25)
    at async generateGraphQLSchema (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\@redwoodjs\internal\dist\generate\graphqlSchema.js:35:15)
    at async generate (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\@redwoodjs\internal\dist\generate\generate.js:16:22)
    at async run (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\@redwoodjs\internal\dist\generate\generate.js:28:17) {
  errors: [
    DetailedError: Failed to load schema
        at loadSchema (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\@graphql-codegen\cli\index.cjs.js:413:15)
        at async C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\@graphql-codegen\cli\index.cjs.js:842:55
        at async Task.task (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\@graphql-codegen\cli\index.cjs.js:696:17) {
      details: '\n' +
        '        Failed to load schema from \n' +
        'scalar Date\n' +
        'scalar Time\n' +
        'scalar DateTime\n' +
        'scalar JSON\n' +
        'scalar JSONObject\n' +
        '\n' +
        'type Redwood {\n' +
        '  version: String\n' +
        '  currentUser: JSON\n' +
        '  prismaVersion: String\n' +
        '}\n' +
        '\n' +
        'type Query {\n' +
        '  redwood: Redwood\n' +
        '},**/*.sdl.{js,ts}:\n' +
        '\n' +
        '        Unknown type: "Bool".\n' +
        '        Error: Unknown type: "Bool".\n' +
        '    at getNamedType (C:\\Users\\tobbe\\dev\\redwood\\acm-store-rw\\node_modules\\graphql\\utilities\\extendSchema.js:360:13)\n' +
        '    at getWrappedType (C:\\Users\\tobbe\\dev\\redwood\\acm-store-rw\\node_modules\\graphql\\utilities\\extendSchema.js:375:12)\n' +
        '    at getWrappedType (C:\\Users\\tobbe\\dev\\redwood\\acm-store-rw\\node_modules\\graphql\\utilities\\extendSchema.js:372:45)\n' +
        '    at buildFieldMap (C:\\Users\\tobbe\\dev\\redwood\\acm-store-rw\\node_modules\\graphql\\utilities\\extendSchema.js:409:17)\n' +
        '    at fields (C:\\Users\\tobbe\\dev\\redwood\\acm-store-rw\\node_modules\\graphql\\utilities\\extendSchema.js:561:22)\n' +
        '    at resolveThunk (C:\\Users\\tobbe\\dev\\redwood\\acm-store-rw\\node_modules\\graphql\\type\\definition.js:480:40)\n' +
        '    at defineFieldMap (C:\\Users\\tobbe\\dev\\redwood\\acm-store-rw\\node_modules\\graphql\\type\\definition.js:692:18)\n' +
        '    at GraphQLObjectType.getFields (C:\\Users\\tobbe\\dev\\redwood\\acm-store-rw\\node_modules\\graphql\\type\\definition.js:633:27)\n' +
        '    at collectReferencedTypes (C:\\Users\\tobbe\\dev\\redwood\\acm-store-rw\\node_modules\\graphql\\type\\schema.js:366:81)\n' +
        '    at new GraphQLSchema (C:\\Users\\tobbe\\dev\\redwood\\acm-store-rw\\node_modules\\graphql\\type\\schema.js:148:9)\n' +
        '    \n' +
        '        GraphQL Code Generator supports:\n' +
        '          - ES Modules and CommonJS exports (export as default or named export "schema")\n' +
        '          - Introspection JSON File\n' +
        '          - URL of GraphQL endpoint\n' +
        '          - Multiple files with type definitions (glob expression)\n' +
        '          - String in config file\n' +
        '    \n' +
        '        Try to use one of above options and run codegen again.\n' +
        '    \n' +
        '      ',
      source: 'C:\\Users\\tobbe\\dev\\redwood\\acm-store-rw\\.redwood\\schema.graphql',
      context: [Object: null prototype] {}
    }
  ],
  context: [Object: null prototype] {}
}
```

In the middle of all that I can see that `Bool` isn't a valid type. I'd like to try to filter out some of the noise.  One way to do that is to only print the error details

```
for (const error of e.errors) {
  console.error('Error details:');
  console.error(error.details)
}
```

Leaving us with this output

```
Error details:

        Failed to load schema from
scalar Date
scalar Time
scalar DateTime
scalar JSON
scalar JSONObject

type Redwood {
  version: String
  currentUser: JSON
  prismaVersion: String
}

type Query {
  redwood: Redwood
},**/*.sdl.{js,ts}:

        Unknown type: "Bool".
        Error: Unknown type: "Bool".
    at getNamedType (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\graphql\utilities\extendSchema.js:360:13)
    at getWrappedType (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\graphql\utilities\extendSchema.js:375:12)
    at getWrappedType (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\graphql\utilities\extendSchema.js:372:45)
    at buildFieldMap (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\graphql\utilities\extendSchema.js:409:17)
    at fields (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\graphql\utilities\extendSchema.js:561:22)
    at resolveThunk (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\graphql\type\definition.js:480:40)
    at defineFieldMap (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\graphql\type\definition.js:692:18)
    at GraphQLObjectType.getFields (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\graphql\type\definition.js:633:27)
    at collectReferencedTypes (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\graphql\type\schema.js:366:81)
    at new GraphQLSchema (C:\Users\tobbe\dev\redwood\acm-store-rw\node_modules\graphql\type\schema.js:148:9)

        GraphQL Code Generator supports:
          - ES Modules and CommonJS exports (export as default or named export "schema")
          - Introspection JSON File
          - URL of GraphQL endpoint
          - Multiple files with type definitions (glob expression)
          - String in config file

        Try to use one of above options and run codegen again.
```

Depending on how much we can/want to trust that the output stays the same we can filter further, by for example removing the useless stack trace, etc.

### PR DESCRIPTION
**Pull request by dac09**
Fixes #2964
Fixes #3351

### What does it do?
1) Print errors on schema generation failure
2) Adds more plugins to babel parser for type generation
3) We used to pass the contents of the file to the ast helper methods in rwjs/internal. This PR changes it so that we now pass the AST to these functions.

**Why?** 
a) So we can display better errors when parsing fails, pin pointing which file it failed on (i.e. the parser is aware of the path, so we can throw the error from the parser)
b) Determine which plugins to use for the parser (the jsx plugin on the TS typecast syntax used to fail)
c) We don't repeatedly pass path to the helper functions just to throw an error

### CODE REVIEW

**Comment by dac09** [comment]
Thanks for the tips @peterp on parser plugins. Please have a sneak peak while I finalise this PR :), and if you agree with this approach! 

I really liked that we could just pass code to the ast helpers (made for nicer tests) - but this seems to me, to be the best approach without reading a file multiple times, or just passing paths for printing errors.

Open to suggestions ✌️

File: packages/internal/package.json [hunk 1]
```diff
@@ -14,10 +14,8 @@
   "dependencies": {
     "@babel/parser": "7.15.3",
     "@babel/plugin-transform-typescript": "7.15.0",
-    "babel-plugin-polyfill-corejs3": "0.2.4",
```
Comment (peterp) [inline packages/internal/package.json:17]: Didn't I just add these packages?

Comment (dac09) [reply packages/internal/package.json:17]: They're duped for some reason!

Comment (dac09) [reply packages/internal/package.json:17]: (see line 29 and 30)

File: packages/internal/src/ast.ts [hunk 2]
```diff
@@ -1,9 +1,42 @@
...
+import { isFileInsideFolder } from './files'
+import { getPaths } from './paths'
+
+export const fileToAst = (filePath: string): types.Node => {
+  const code = fs.readFileSync(filePath, 'utf-8')
 
-export const parse = (code: string) =>
-  babelParse(code, { sourceType: 'module', plugins: ['jsx', 'typescript'] })
+  const plugins: ParserPlugin[] = [
+    'typescript',
```
Comment (peterp) [inline packages/internal/src/ast.ts:16]: you could also turn off "typescript" if it's not a ts/tsx file.

Comment (peterp) [reply packages/internal/src/ast.ts:16]: The `.filter(Boolean)` pattern may be helpful here.

**Review by peterp — APPROVED** [review]
Looks good to me.

I left a comment about the removal of those babel polyfill packages. I'm sure they're pulled in via other packages, but I would rather be explicit about including them.

File: packages/internal/package.json (hunk 1 above)
Comment (peterp) [reply packages/internal/package.json:17]: Oh. The brain only parses the red and green lines.

**Comment by peterp** [comment]
Super nice approach Danny 👍

File: packages/internal/src/ast.ts (hunk 2 above)
Comment (peterp) [reply packages/internal/src/ast.ts:16]: 👍

**Pull request merged**